File: q0966/q0966.py
class Solution:
    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        res = []

        wordSet = set(wordlist)
        lower_map = dict()
        vowel_map = dict()

        for word in wordlist:
            lower = word.lower()
            devow = self.devowel(word)

            if lower not in lower_map:
                lower_map[lower] = word
            if devow not in vowel_map:
                vowel_map[devow] = word

        for q in queries:
            if q in wordSet:
                res.append(q)
            elif q.lower() in lower_map:
                res.append(lower_map[q.lower()])
            elif self.devowel(q) in vowel_map:
                res.append(vowel_map[self.devowel(q)])
            else:
                res.append("")

        return res

    def devowel(self, word: str) -> str:
        return ''.join('*' if c.lower() in 'aeiou' else c.lower() for c in word)

File: q0966/q0966.java
class Solution {
    Set<String> wordSet = new HashSet<>();
    Map<String, String> wordCap = new HashMap<>();
    Map<String, String> wordVow = new HashMap<>();

    public String[] spellchecker(String[] wordlist, String[] queries) {
        for (String word : wordlist) {
            wordSet.add(word);
            
            String wordLower = word.toLowerCase();
            wordCap.putIfAbsent(wordLower, word);

            String wordDevowel = devowel(word);
            wordVow.putIfAbsent(wordDevowel, word);
        }

        String[] res = new String[queries.length];
        for (int i = 0; i < queries.length; i++) {
            final String solvedQuery = solve(queries[i]);
            res[i] = solvedQuery;
        }
        return res;
    }

    private String solve(String query) {
        if (wordSet.contains(query)) return query;
        
        String queryLower = query.toLowerCase(); 
        if (wordCap.containsKey(queryLower)) return wordCap.get(queryLower);

        String queryDevowel = devowel(query);
        if (wordVow.containsKey(queryDevowel)) return wordVow.get(queryDevowel);

        return "";
    }

    private String devowel(String s) {
        s = s.toLowerCase();  // Ensure consistent casing
        StringBuilder sb = new StringBuilder();
        for (char c : s.toCharArray()) {
            if ("aeiou".indexOf(c) == -1) {
                sb.append(c);
            } else {
                sb.append('*');
            }
        }
        return sb.toString();
    }
}
